src/jobserver_test.cc

```diff
@@ -0,0 +1,345 @@
+
+  /// Return the number of tokens currently available in the external token
+  /// pool.
+  size_t ServerCount();
+
+  /// Remove/close the handle to external token pool.
+  void ServerRemove();
+
+  /// Wrapped jobserver object to test on.
+  JobserverTest jobserver_;
+
+  /// Stored makeflags read before starting tests.
+  const char* makeflags_ = nullptr;
+
+  /// Name of created external jobserver token pool.
+  const char* name_ = nullptr;
+
+#ifdef _WIN32
+  /// Implementation of posix setenv() for windows that forwards calls to
+  /// _putenv().
+  int setenv(const char* name, const char* value, int _) {
+    std::string envstring;
+
+    // _putenv() requires a single <name>=<value> string as argument.
+    envstring += name;
+    envstring += '=';
+    envstring += value;
+
+    return _putenv(envstring.c_str());
+  };
+
+  /// Implementation of posix unsetenv() for windows that forwards calls to
+  /// _putenv().
+  int unsetenv(const char* name) {
+    /// Call _putenv() with <name>="" to unset the env variable.
+    return setenv(name, "", 0);
+  }
+
+  /// Handle of the semaphore used as external token pool.
+  HANDLE sem_ = INVALID_HANDLE_VALUE;
+#else
+  /// File descriptor of the fifo used as external token pool.
+  int fd_ = -1;
+#endif
+};
+
+JobserverTestState::JobserverTestState() {
+  makeflags_ = getenv("MAKEFLAGS");
+  unsetenv("MAKEFLAGS");
+}
+
+JobserverTestState::~JobserverTestState() {
+  if (name_ != nullptr) {
+    ServerRemove();
+  }
+
+  if (makeflags_ != nullptr) {
+    setenv("MAKEFLAGS", makeflags_, 1);
+  } else {
+    unsetenv("MAKEFLAGS");
+  }
+}
+
+void JobserverTestState::ServerConfigure(const char* name)
+{
+  std::string makeflags("--jobserver-auth=");
+#ifdef _WIN32
+  makeflags += "sem:";
+#else
+  makeflags += "fifo:";
+#endif
+  makeflags += name;
+
+  assert(setenv("MAKEFLAGS", makeflags.c_str(), 1) == 0);
+
+}
+
+#ifdef _WIN32
+void JobserverTestState::ServerCreate(const char* name, size_t count) {
+  assert(name_ == nullptr);
+  ServerConfigure(name);
+  name_ = name;
+
+  // One cannot create a semaphore with a max value of 0 on windows
+  sem_ = CreateSemaphoreA(nullptr, count, count ? count : 1, name);
+  assert(sem_ != nullptr);
+}
+
+size_t JobserverTestState::ServerCount() {
+  assert(name_ != nullptr);
+  size_t count = 0;
+
+  // First acquire all the available tokens to count them
+  while (WaitForSingleObject(sem_, 0) == WAIT_OBJECT_0) {
+    count++;
+  }
+
+  // Then return the acquired tokens to let the client continue
+  ReleaseSemaphore(sem_, count, nullptr);
+
+  return count;
+}
+
+void JobserverTestState::ServerRemove() {
+  assert(name_ != nullptr);
+  CloseHandle(sem_);
+  name_ = nullptr;
+}
+
+#else // _WIN32
+
+void JobserverTestState::ServerCreate(const char* name, size_t count) {
+  assert(name_ == nullptr);
+  ServerConfigure(name);
+  name_ = name;
+
+  // Create and open the fifo
+  assert(mkfifo(name, S_IWUSR | S_IRUSR) == 0);
```

Comment:
I recommend removing the file if it already exists, as this can happen if your unit-test crashed previously, making this `mkfifo()` call fail with EEXIST.